File: mobile/packages/wms_api_client/lib/src/dto/master_data/master_data_dtos.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:wms_core/wms_core.dart';

import '../../json/date_only_converter.dart';

part 'master_data_dtos.freezed.dart';
part 'master_data_dtos.g.dart';

/// `master_uom`.
@freezed
abstract class UomDto with _$UomDto {
  const factory UomDto({
    required int id,
    required String code,
    required String name,
    required UomClass uomClass,
    @Default(3) int decimals,
  }) = _UomDto;

  factory UomDto.fromJson(Map<String, Object?> json) => _$UomDtoFromJson(json);
}

/// `master_product_category`.
@freezed
abstract class ProductCategoryDto with _$ProductCategoryDto {
  const factory ProductCategoryDto({
    required int id,
    required String code,
    required String name,
    required ProductType productType,
    required String path,
    int? parentId,
  }) = _ProductCategoryDto;

  factory ProductCategoryDto.fromJson(Map<String, Object?> json) =>
      _$ProductCategoryDtoFromJson(json);
}

/// `master_product_uom` - alternative units with a frozen factor.
@freezed
abstract class ProductUomDto with _$ProductUomDto {
  const factory ProductUomDto({
    required int id,
    required int uomId,
    required Decimal factorToBase,
    @DateOnlyConverter() required DateTime validFrom,
    String? uomCode,
    @Default(false) bool isPurchaseDefault,
    @Default(false) bool isIssueDefault,
    @NullableDateOnlyConverter() DateTime? validTo,
  }) = _ProductUomDto;

  factory ProductUomDto.fromJson(Map<String, Object?> json) =>
      _$ProductUomDtoFromJson(json);
}

/// `master_product`.
///
/// Cost fields ([avgUnitCost], [lastPurchasePrice]) are **absent** from the
/// JSON for users without `master.product.view_cost` (spec §16); they are
/// therefore nullable and the UI must hide them when `null`.
@freezed
abstract class ProductDto with _$ProductDto {
  const factory ProductDto({
    required int id,
    required String sku,
    required String name,
    required int categoryId,
    required int baseUomId,
    required Decimal vatRate,
    String? barcode,
    String? categoryName,
    String? brand,
    String? baseUomCode,
    int? defaultSupplierId,
    Quantity? minStock,
    Quantity? maxStock,
    Quantity? reorderPoint,
    @Default(false) bool requiresBatch,
    @Default(false) bool requiresExpiry,
    @Default(IssueStrategy.fefo) IssueStrategy issueStrategy,
    int? shelfLifeDays,
    String? imageKey,
    @Default(true) bool isActive,
    @Default(1) int rowVersion,
    @Default(<ProductUomDto>[]) List<ProductUomDto> uoms,
    Money? avgUnitCost,
    Money? lastPurchasePrice,
    String? currency,
  }) = _ProductDto;

  const ProductDto._();

  factory ProductDto.fromJson(Map<String, Object?> json) =>
      _$ProductDtoFromJson(json);

  /// `true` when the server included cost information.
  bool get hasCostInfo => avgUnitCost != null || lastPurchasePrice != null;
}

/// `master_supplier`.
@freezed
abstract class SupplierDto with _$SupplierDto {
  const factory SupplierDto({
    required int id,
    required String code,
    required String name,
    @Default('AZN') String currency,
    String? taxId,
    String? contactPerson,
    String? phone,
    String? email,
    String? address,
    String? paymentTerms,
    String? deliveryTerms,
    String? incoterms,
    @Default(false) bool isApprovedFoodSupplier,
    @Default(true) bool isActive,
    @Default(1) int rowVersion,
  }) = _SupplierDto;

  factory SupplierDto.fromJson(Map<String, Object?> json) =>
      _$SupplierDtoFromJson(json);
}

/// `master_location` - physical and virtual.
@freezed
abstract class LocationDto with _$LocationDto {
  const factory LocationDto({
    required int id,
    required String code,
    required String name,
    required LocationType locationType,
    required bool isVirtual,
    int? parentId,
    @Default(true) bool allowsFood,
    @Default(true) bool allowsNonFood,
    @Default(true) bool isActive,
  }) = _LocationDto;

  factory LocationDto.fromJson(Map<String, Object?> json) =>
      _$LocationDtoFromJson(json);
}

/// `master_reason_code`.
@freezed
abstract class ReasonCodeDto with _$ReasonCodeDto {
  const factory ReasonCodeDto({
    required int id,
    required String code,
    required String name,
    required ReasonGroup reasonGroup,
    @Default(true) bool requiresApproval,
    @Default(false) bool requiresPhoto,
    @Default(true) bool isActive,
  }) = _ReasonCodeDto;

  factory ReasonCodeDto.fromJson(Map<String, Object?> json) =>
      _$ReasonCodeDtoFromJson(json);
}

File: backend/src/Modules/MasterData/Wms.MasterData.Application/Dtos/MasterDataDtos.cs
using Wms.MasterData.Domain.Enums;

namespace Wms.MasterData.Application.Dtos;

/// <summary><c>common.v1.yaml#/components/schemas/AuditFields</c> (spec §6.2).</summary>
public sealed record AuditFieldsDto(
    DateTimeOffset CreatedAt,
    uint CreatedBy,
    DateTimeOffset? UpdatedAt,
    uint? UpdatedBy,
    uint RowVersion);

// ==================================================================== products

/// <summary><c>ProductSummary</c> — the list row and the shape other screens embed.</summary>
public sealed record ProductSummaryDto(
    uint Id,
    string Sku,
    string Name,
    string? Barcode,
    ushort BaseUomId,
    string BaseUomCode,
    ProductType ProductType,
    bool RequiresBatch,
    bool RequiresExpiry,
    bool IsActive);

/// <summary><c>ProductUom</c> — one versioned conversion factor row (spec §12.1).</summary>
public sealed record ProductUomDto(
    uint Id,
    uint ProductId,
    ushort UomId,
    string UomCode,
    decimal FactorToBase,
    bool IsPurchaseDefault,
    bool IsIssueDefault,
    DateOnly ValidFrom,
    DateOnly? ValidTo);

/// <summary><c>Product</c> — the full card returned by get/create/update.</summary>
public sealed record ProductDetailDto(
    uint Id,
    string Sku,
    string Name,
    string? Barcode,
    ushort BaseUomId,
    string BaseUomCode,
    ProductType ProductType,
    bool RequiresBatch,
    bool RequiresExpiry,
    bool IsActive,
    uint CategoryId,
    string CategoryPath,
    string? Brand,
    uint? DefaultSupplierId,
    decimal? MinStock,
    decimal? MaxStock,
    decimal? ReorderPoint,
    decimal VatRate,
    IssueStrategy IssueStrategy,
    ushort? ShelfLifeDays,
    long? ImageAttachmentId,
    IReadOnlyList<ProductUomDto> Uoms,
    AuditFieldsDto Audit);

// ==================================================================== categories

public sealed record CategoryDto(
    uint Id,
    uint? ParentId,
    string Code,
    string Name,
    ProductType ProductType,
    string Path,
    IssueStrategy? DefaultIssueStrategy,
    bool IsActive,
    uint RowVersion);

// ==================================================================== units of measure

public sealed record UomDto(
    ushort Id,
    string Code,
    string Name,
    UomClass UomClass,
    byte Decimals);

// ==================================================================== suppliers

public sealed record SupplierSummaryDto(
    uint Id,
    string Code,
    string Name,
    string? TaxId,
    string Currency,
    bool IsApprovedFoodSupplier,
    bool IsActive);

public sealed record SupplierCertificateDto(
    uint Id,
    uint SupplierId,
    string CertType,
    string? CertNumber,
    DateOnly? IssuedDate,
    DateOnly? ExpiryDate,
    long? AttachmentId,
    bool IsExpired);

public sealed record SupplierDetailDto(
    uint Id,
    string Code,
    string Name,
    string? TaxId,
    string Currency,
    bool IsApprovedFoodSupplier,
    bool IsActive,
    string? ContactPerson,
    string? Phone,
    string? Email,
    string? Address,
    string? BankDetails,
    string? PaymentTerms,
    string? DeliveryTerms,
    string? Incoterms,
    IReadOnlyList<SupplierCertificateDto> Certificates,
    AuditFieldsDto Audit);

// ==================================================================== locations

public sealed record LocationDetailDto(
    uint Id,
    uint? ParentId,
    string Code,
    string Name,
    LocationType LocationType,
    bool IsVirtual,
    bool AllowsFood,
    bool AllowsNonFood,
    bool IsActive,
    uint RowVersion);

// ==================================================================== currency rates

public sealed record CurrencyRateDto(
    uint Id,
    string Currency,
    DateOnly RateDate,
    decimal RateToBase,
    string Source);

// ==================================================================== reason codes

public sealed record ReasonCodeDto(
    ushort Id,
    string Code,
    string Name,
    ReasonGroup ReasonGroup,
    bool RequiresApproval,
    bool RequiresPhoto,
    bool IsActive,
    uint RowVersion);

// ==================================================================== number sequences

public sealed record NumberSequenceDto(
    uint Id,
    string DocType,
    string Prefix,
    string Period,
    uint LastNumber,
    byte Padding,
    string NextNumberPreview);
